File: go/bottle-song/bottle_song.go
package bottlesong

import (
	"fmt"
	"unicode"
)

// Recite generates a series of verses from the "Ten Green Bottles" song.
// It starts with a specified number of bottles and recites a given number of verses.
func Recite(startBottles, takeDown int) []string {
	verses := []string{}

	for i := 0; i < takeDown; i++ {
		verses = append(verses, verse(startBottles-i)...)
		if i < takeDown-1 {
			verses = append(verses, "")
		}
	}

	return verses
}

// verse generates a single four-line verse of the song for a given number of bottles.
func verse(hangingCount int) []string {
	const lineThree = "And if one green bottle should accidentally fall,"

	lineOneAndTwo := bottlesHangingOnTheWall(hangingCount)
	lineFour := thereWillBeHanging(hangingCount - 1)

	return []string{lineOneAndTwo, lineOneAndTwo, lineThree, lineFour}
}

// bottlesHangingOnTheWall formats the first two lines of a verse.
// It capitalizes the quantity word and handles the pluralization of "bottle".
func bottlesHangingOnTheWall(count int) string {
	const lineFormat = "%s green bottle%s hanging on the wall,"

	sayCount := capitalize(quantities[count])
	terminatorForBottle := tern(count == 1, "", "s")

	return fmt.Sprintf(lineFormat, sayCount, terminatorForBottle)
}

// thereWillBeHanging formats the last line of a verse.
// It states how many bottles are left and handles the pluralization.
func thereWillBeHanging(count int) string {
	const format = "There'll be %s green bottle%s hanging on the wall."

	sayCount := quantities[count]
	terminator := tern(count == 1, "", "s")

	return fmt.Sprintf(format, sayCount, terminator)
}

// tern is a helper function that acts as a simple ternary operator.
// It returns trueCase if the condition is true, otherwise it returns falseCase.
func tern(cond bool, trueCase string, falseCase string) string {
	if cond {
		return trueCase
	} else {
		return falseCase
	}
}

// capitalize is a helper function that capitalizes the first letter of a word.
func capitalize(word string) string {
	runes := []rune(word)
	runes[0] = unicode.ToUpper(runes[0])
	return string(runes)
}

// quantities is a slice that maps integer bottle counts to their corresponding word.
var quantities = []string{
	"no", "one", "two", "three", "four",
	"five", "six", "seven", "eight", "nine", "ten",
}

File: lua/bottle-song/bottle-song.lua
--- A module that generates the "Green Bottles" song.
local BottleSong = {}

local verse, capitalize, terminator_for

--- Recites a specified number of verses of the "Green Bottles" song.
--- @param start_bottles integer The starting number of bottles.
--- @param take_down integer The number of verses to generate.
--- @return string The generated verses, separated by newlines.
function BottleSong.recite(start_bottles, take_down)
    local verses = {}
    local haging_on_last_verse = start_bottles - take_down + 1

    for hanging = start_bottles, haging_on_last_verse, -1 do
        verses[#verses + 1] = verse(hanging)
    end

    return table.concat(verses, "\n")
end

--- The format string for a single verse of the song.
local VERSE_FORMAT <const> = [[
%s green bottle%s hanging on the wall,
%s green bottle%s hanging on the wall,
And if one green bottle should accidentally fall,
There'll be %s green bottle%s hanging on the wall.
]];

--- A constant table mapping bottle counts to their string representations.
local SAID_COUNTS <const> = {
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
    "ten",
    [0] = "no"
}

--- Generates a single verse of the song based on the number of bottles hanging.
--- @param hanging_count integer The number of bottles for this verse.
--- @return string The formatted verse.
function verse(hanging_count)
    local initial_count_said = capitalize(SAID_COUNTS[hanging_count])
    local initial_terminator = terminator_for(hanging_count)
    local next_count = SAID_COUNTS[hanging_count - 1]
    local next_terminator = terminator_for(hanging_count - 1)

    return VERSE_FORMAT:format(
        initial_count_said, initial_terminator,
        initial_count_said, initial_terminator,
        next_count, next_terminator
    )
end

--- Capitalizes the first letter of a string.
--- @param str string The string to capitalize.
--- @return string The capitalized string.
function capitalize(str)
    return str:sub(1, 1):upper() .. str:sub(2)
end

--- Returns the plural terminator "s" for a count, or an empty string for a singular count.
--- @param count integer The bottle count.
--- @return string The plural terminator ("s") or an empty string.
function terminator_for(count)
    if count == 1 then
        return ""
    else
        return "s"
    end
end

return BottleSong